modules/navigation/godot_navigation_server.cpp

```diff
@@ -808,8 +814,117 @@ void GodotNavigationServer::process(real_t p_delta_time) {
+	ERR_FAIL_COND_V(map == nullptr, r_query_result);
+
+	// run the pathfinding
+
+	if (p_parameters.pathfinding_algorithm == gd::PathfindingAlgorithm::PATHFINDING_ALGORITHM_ASTAR) {
+		r_query_result.path = map->get_path(p_parameters.start_position, p_parameters.target_position, p_parameters.optimize, p_parameters.navigation_layers);
+	} else {
+		return r_query_result;
+	}
+
+	// add path postprocessing
+
+	// add calculated path stats
+
+	if (r_query_result.path.size() > 1 && p_parameters.calculate_path_length) {
+		_calculate_path_length(p_parameters, r_query_result);
+	}
+
+	return r_query_result;
+}
+
+void GodotNavigationServer::_calculate_path_length(const gd::PathQueryParameters &p_parameters, gd::PathQueryResult &p_query_result) const {
+	float _sum = 0.0;
+	if (p_query_result.path.size() > 1) {
+		for (int i = 1; i < p_query_result.path.size(); i++) {
+			const Vector3 &prev_path_pos = p_query_result.path[i - 1];
+			const Vector3 &path_pos = p_query_result.path[i];
+			_sum += prev_path_pos.distance_to(path_pos);
+		}
+	}
+	p_query_result.path_length = _sum;
+}
+
+void GodotNavigationServer::_process_path_query(gd::PathQueryTask *pathquery_task) {
+	pathquery_task->processed = true;
+
+	Object *queryresult_object = ObjectDB::get_instance(pathquery_task->result_object_id);
+	if (!queryresult_object) {
+		// query result object id empty, nothing to call home
+		return;
+	}
+
+	const gd::PathQueryResult _query_result = _query_path(pathquery_task->query_parameters);
+
+	// NavigationPathQueryResult2D
+	if (Object::cast_to<NavigationPathQueryResult2D>(queryresult_object) != nullptr) {
+		NavigationPathQueryResult2D *pathqueryresult2d = Object::cast_to<NavigationPathQueryResult2D>(queryresult_object);
+
+		Vector<Vector2> path_vec2;
+		path_vec2.resize(_query_result.path.size());
+		for (int i = 0; i < _query_result.path.size(); i++) {
+			path_vec2.set(i, Vector2(_query_result.path[i].x, _query_result.path[i].z));
+		}
+		pathqueryresult2d->set_path(path_vec2);
+		pathqueryresult2d->set_path_length(_query_result.path_length);
+
+		// NavigationPathQueryResult3D
+	} else if (Object::cast_to<NavigationPathQueryResult3D>(queryresult_object) != nullptr) {
+		NavigationPathQueryResult3D *pathqueryresult3d = Object::cast_to<NavigationPathQueryResult3D>(queryresult_object);
+
+		pathqueryresult3d->set_path(_query_result.path);
+		pathqueryresult3d->set_path_length(_query_result.path_length);
+
+	} else {
+		// no valid result object, no need for callback
+		return;
+	}
+
+	Object *query_callback_object = ObjectDB::get_instance(pathquery_task->query_parameters.callback_object_id);
+	StringName query_callback_method = pathquery_task->query_parameters.callback_method;
+
+	if (query_callback_object && query_callback_method != StringName()) {
+		const Variant **args = {};
+		int arg_count = 0;
+		Callable::CallError callback_error;
+		// query callback object
+		query_callback_object->callp(query_callback_method, args, arg_count, callback_error);
+	}
```

Comment:
Are there any race conditions here? Can the `query_callback_object` be deleted from another thread in between the time that you get it and call `callp`?

Reply:
As far as I am aware of the Godot loop in the time window that the NavigationServer process this there is no deletion going on as all the user scripts frees() or queue_free() deletions of objects are already handled for the frame way before? Does it still require a lock in this case?

Reply:
removed from this pr